File: leetcode/20.ValidParentheses.go
package leetcode

func isValid(s string) bool {
	var stack []string
	obj := map[string]string{
		"(": ")",
		"{": "}",
		"[": "]",
	}

	for _, char := range s {
		mapped, ok := obj[string(char)]
		if ok {
			stack = append(stack, mapped)
			continue
		}
		if len(stack) > 0 && stack[len(stack)-1] == string(char) {
			stack = stack[:len(stack)-1]
			continue
		}

		return false
	}

	return len(stack) == 0
}

File: leetcode/20.ValidParentheses.ts
function isValid(s: string): boolean {
  const map = {
    '(': ')',
    '{': '}',
    '[': ']',
  } as const
  type Key = keyof typeof map

  const stack = []

  for (const element of s) {
    if (map[element as Key]) {
      stack.push(map[element as Key])
      continue
    }
    if (stack[stack.length - 1] === element) {
      stack.pop()
      continue
    }

    return false
  }

  return stack.length === 0
}
console.log(isValid('({{{{}}}})')) // false
